File: backend/tools/write_file.py
import logging
from pathlib import Path

from backend.core.file_access import (
    atomic_write_bytes,
    atomic_write_text,
    backup_if_exists,
    is_write_allowed,
    resolve_path,
)
from backend.core.python_syntax import is_python_source_file
from backend.core.source_syntax import is_syntax_checked_file, validate_source_syntax

logger = logging.getLogger(__name__)

TOOL_DEF = {
    "name": "write_file",
    "description": "Write content to a file. Use relative paths from the project root (e.g. backend/tools/read_file.py). Paths are canonicalized and must stay inside the project. Overwrites are done atomically (tmp then replace). Overwrites by default; set append to true to add at the end. Parent directories are created if missing. Read the file first when editing so you don't overwrite blindly. Writes to .git/ and node_modules/ are blocked.",
    "input_schema": {
        "type": "object",
        "properties": {
            "file_path": {
                "type": "string",
                "description": "Path to the file (relative to project root or absolute within project)",
            },
            "content": {
                "type": "string",
                "description": "Content to write (hex string if mode is 'binary')",
            },
            "mode": {
                "type": "string",
                "description": "File mode: 'text' or 'binary' (default: text)",
            },
            "append": {
                "type": "boolean",
                "description": "If true, append to the file instead of overwriting (default: false)",
            },
            "backup": {
                "type": "boolean",
                "description": "If true and overwriting an existing file, create a .bak copy first (default: false)",
            },
            "encoding": {
                "type": "string",
                "description": "Text encoding for text mode (default: utf-8)",
            },
        },
        "required": ["file_path", "content"],
    },
    "default_enabled": True,
}


def call_tool(tool_input: dict, context=None):
    file_path = tool_input.get("file_path", "").strip()
    content = tool_input.get("content", "")
    mode = (tool_input.get("mode") or "text").lower()
    append = tool_input.get("append", False)
    backup = tool_input.get("backup", False)
    encoding = tool_input.get("encoding") or "utf-8"

    resolved = resolve_path(file_path)
    if resolved is None:
        logger.warning("Rejected path outside workspace: %s", file_path)
        return 200, {
            "success": False,
            "error": "Path is outside the project workspace.",
            "path": file_path,
        }

    allowed, err = is_write_allowed(resolved)
    if not allowed:
        logger.warning("Write blocked for %s: %s", file_path, err)
        return 200, {"success": False, "error": err, "path": str(resolved)}

    path_str = str(resolved)
    if resolved.exists() and resolved.is_dir():
        return 200, {"success": False, "error": "Path is a directory, not a file.", "path": path_str}

    try:
        resolved.parent.mkdir(parents=True, exist_ok=True)

        if mode == "text" and is_syntax_checked_file(resolved):
            if append and resolved.exists():
                try:
                    to_check = resolved.read_text(encoding=encoding) + content
                except UnicodeDecodeError as ude:
                    return 200, {
                        "success": False,
                        "error": f"Cannot validate source: file is not valid {encoding}: {ude}",
                        "path": path_str,
                    }
            else:
                to_check = content
            syn_err = validate_source_syntax(resolved, to_check)
            if syn_err:
                return 200, {"success": False, "error": syn_err, "path": path_str}

        if mode == "text":
            if append:
                with open(resolved, "a", encoding=encoding) as f:
                    f.write(content)
            else:
                if backup and resolved.exists():
                    backup_if_exists(resolved)
                atomic_write_text(resolved, content, encoding=encoding)
        elif mode == "binary":
            data = bytes.fromhex(content)
            if append:
                with open(resolved, "ab") as f:
                    f.write(data)
            else:
                if backup and resolved.exists():
                    backup_if_exists(resolved)
                atomic_write_bytes(resolved, data)
        else:
            return 200, {
                "success": False,
                "error": f"Invalid mode: {mode}. Use 'text' or 'binary'.",
                "path": path_str,
            }

        size = resolved.stat().st_size
        logger.info("Successfully wrote to file: %s", path_str)
        out = {
            "success": True,
            "path": path_str,
            "size": size,
            "filename": resolved.name,
            "append": append,
        }
        if mode == "text" and is_syntax_checked_file(resolved):
            out["syntax_ok"] = True
        if mode == "text" and is_python_source_file(resolved):
            out["python_syntax_ok"] = True
        return 200, out
    except Exception as e:
        logger.error("Error writing to file %s: %s", path_str, e, exc_info=True)
        return 200, {"success": False, "error": str(e), "path": path_str}

File: akira-desktop/electron/tools/file-operations/write-file.js
/**
 * Write File Tool
 * write_file - Write content to a file
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

// Workspace root - default to user's home directory
let WORKSPACE_ROOT = os.homedir();

function setWorkspaceRoot(root) {
  WORKSPACE_ROOT = root;
}

/**
 * Resolve and validate path stays within workspace
 */
function resolvePath(filePath) {
  if (!filePath) return null;

  // Handle absolute paths
  let resolved;
  if (path.isAbsolute(filePath)) {
    resolved = path.normalize(filePath);
  } else {
    resolved = path.normalize(path.join(WORKSPACE_ROOT, filePath));
  }

  return resolved;
}

/**
 * Check if write is allowed (block certain paths)
 */
function isWriteAllowed(filePath) {
  const blocked = ['.git', 'node_modules', 'System32', 'Windows'];
  const pathLower = filePath.toLowerCase();

  for (const b of blocked) {
    if (pathLower.includes(path.sep + b.toLowerCase() + path.sep) ||
        pathLower.endsWith(path.sep + b.toLowerCase())) {
      return { allowed: false, error: `Writing to ${b} is blocked for safety.` };
    }
  }
  return { allowed: true };
}

const definitions = [
  {
    name: 'write_file',
    description: 'Write content to a file. Creates parent directories if missing. Use append mode to add to existing file. Writes to .git/ and node_modules/ are blocked.',
    input_schema: {
      type: 'object',
      properties: {
        file_path: {
          type: 'string',
          description: 'Path to the file (relative to workspace or absolute)',
        },
        content: {
          type: 'string',
          description: 'Content to write (hex string if mode is "binary")',
        },
        mode: {
          type: 'string',
          description: 'File mode: "text" or "binary" (default: text)',
        },
        append: {
          type: 'boolean',
          description: 'If true, append to file instead of overwriting (default: false)',
        },
        backup: {
          type: 'boolean',
          description: 'If true and overwriting existing file, create a .bak copy first (default: false)',
        },
        encoding: {
          type: 'string',
          description: 'Text encoding for text mode (default: utf-8)',
        },
      },
      required: ['file_path', 'content'],
    },
  },
];

const handlers = {
  async write_file(input) {
    const filePath = resolvePath(input.file_path);
    if (!filePath) {
      return { success: false, error: 'Invalid file path' };
    }

    const { allowed, error } = isWriteAllowed(filePath);
    if (!allowed) {
      return { success: false, error };
    }

    const content = input.content || '';
    const mode = (input.mode || 'text').toLowerCase();
    const append = input.append || false;
    const backup = input.backup || false;
    const encoding = input.encoding || 'utf-8';

    // Validate mode
    if (mode !== 'text' && mode !== 'binary') {
      return { success: false, error: `Invalid mode: ${mode}. Use 'text' or 'binary'.`, path: filePath };
    }

    try {
      // Create parent directories
      const dir = path.dirname(filePath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }

      // Create backup if requested and file exists
      if (backup && !append && fs.existsSync(filePath)) {
        const backupPath = filePath + '.bak';
        fs.copyFileSync(filePath, backupPath);
      }

      if (mode === 'text') {
        if (append) {
          fs.appendFileSync(filePath, content, encoding);
        } else {
          fs.writeFileSync(filePath, content, encoding);
        }
      } else {
        // Binary mode: content is hex string
        const data = Buffer.from(content, 'hex');
        if (append) {
          fs.appendFileSync(filePath, data);
        } else {
          fs.writeFileSync(filePath, data);
        }
      }

      const stats = fs.statSync(filePath);
      return {
        success: true,
        path: filePath,
        size: stats.size,
        filename: path.basename(filePath),
        append,
        mode,
      };
    } catch (error) {
      return { success: false, error: error.message, path: filePath };
    }
  },
};

module.exports = { definitions, handlers, setWorkspaceRoot };
